libsolidity/analysis/ControlFlowGraph.h

```diff
@@ -0,0 +1,150 @@
+	std::vector<CFGNode*> exits;
+
+	/// Control flow in the node.
+	ControlFlowBlock block;
+};
+
+/** Describes the control flow of a function. */
+struct FunctionFlow
+{
+	FunctionFlow(CFGNode* _entry, CFGNode* _exit, CFGNode* _revert):
+		entry(_entry), exit(_exit), revert(_revert) {}
+	virtual ~FunctionFlow() {}
+	/// Entry node. Control flow of the function starts here.
+	/// This node is empty and does not have any entries.
+	CFGNode* entry;
+	/// Exit node. All non-reverting control flow of the function ends here.
+	/// This node is empty and does not have any exits, but may have multiple entries
+	/// (e.g. all return statements of the function).
+	CFGNode* exit;
+	/// Revert node. Control flow of the function in case of revert.
+	/// This node is empty does not have any exists, but may have multiple entries
+	/// (e.g. all assert, require, revert and throw statements).
+	CFGNode* revert;
+};
+
+/** Describes the control flow of a modifier.
+ * Every placeholder breaks the control flow. The node preceding the
+ * placeholder is assigned placeholderEntry as exit and the node
+ * following the placeholder is assigned placeholderExit as entry.
+ */
+struct ModifierFlow: FunctionFlow
+{
+	template<typename... Args>
+	ModifierFlow(Args&&... args): FunctionFlow(std::forward<Args>(args)...) {}
+
+	/// Control flow leading towards a placeholder exit in placeholderEntry.
+	CFGNode* placeholderEntry = nullptr;
+	/// Control flow coming from a placeholder enter from placeholderExit.
+	CFGNode* placeholderExit = nullptr;
+};
+
+class CFG: private ASTConstVisitor
+{
+public:
+	explicit CFG(ErrorReporter& _errorReporter): m_errorReporter(_errorReporter) {}
+
+	bool constructFlow(ASTNode const& _astRoot);
+
+	virtual bool visit(ModifierDefinition const& _modifier) override;
+	virtual bool visit(FunctionDefinition const& _function) override;
+
+	FunctionFlow const& functionFlow(FunctionDefinition const& _function) const;
+private:
+	CFGNode* newNode();
+
+	/// Initially the control flow for all functions *without* modifiers and for
+	/// all modifiers is constructed. Afterwards the control flow of functions
+	/// is adjusted by applying all modifiers.
+	void applyModifiers();
+
+	void applyModifierFlowToFunctionFlow(
+		ModifierFlow const& _modifierFlow,
+		std::shared_ptr<FunctionFlow> _functionFlow
+	);
+
+	ErrorReporter& m_errorReporter;
+
+	std::map<FunctionDefinition const*, std::shared_ptr<FunctionFlow>> m_functionControlFlow;
+	std::map<ModifierDefinition const*, std::shared_ptr<ModifierFlow>> m_modifierControlFlow;
+
+	/// List of nodes.
+	/// All nodes allocated during the construction of the control flow graph
+	/// are owned by the CFG class and stored here.
+	std::vector<std::unique_ptr<CFGNode>> m_nodes;
+
+	/// The control flow of the function that is currently parsed.
+	/// Note: this can also be a ModifierFlow
+	std::shared_ptr<FunctionFlow> m_currentFunctionFlow;
```

Comment:
Why does the CFG own these two variables in the first place? Wouldn't it be better to move them to the parser?

Reply:
That's true, I'll first take care of everything else and then think about this. In any case, unfortunately, I still think it makes sense for ``CFG`` to at least own all nodes and that ``CFG::newNode`` shouldn't be public, so we'd still need the ``friend`` declaration.

Reply:
If these are the only ones, then it might make sense to create a separate class `NodeContainer` that owns the nodes and has a `newNode` function and is a member of CFG, but is passed on to the parser via reference.

Reply:
(private member)

Reply:
Yes, right, this might be the cleanest solution. I'm working on it.